lep/DRAFT/010_Updating_Lantern_Without_User_Intervention.md

```diff
@@ -0,0 +1,68 @@
+
+To ensure users are always using the latest version of Lantern under a
+zero-configuration approach it is necessary to have a mechanism that allows
+Lantern to update itself without user intervention.
+
+## Proposal
+
+Clients can update themselves by downloading the lantern binary and replacing
+the currently installed executable with the new one. They can do this either by
+downloading the full binary or an incremental patch. If the client downloads an
+incremental patch instead of a full binary the bandwidth cost of a single
+client updating will get reduced too.
+
+In order for a client to get access to incremental patches those incremental
+patches must be generated by a third party program over the network, an update
+server.
+
+The Lantern team already uses Github for publishing version releases, the
+update server could use the Github API to get access to such releases, store
+them, compute binary differences between releases and provide public URLs for
+clients to download the patches.
+
+## Implementation details
+
+### General flow
+
+![lanternautoupdates - general client](https://cloud.githubusercontent.com/assets/385670/6097030/736614c8-af72-11e4-932f-07f718c51673.png)
+
+At some point on the Lantern application's lifetime, an independent process
+will be created, this process will periodically send local information (using a
+proxy) to a specially crafted server (the update server) that will compare
+client's data against a list of releases. When applicable, the server will
+generate a binary patch and send a reply to the client containing the URL of
+the appropriate patch. The client will download and apply the patch to its
+executable file so the new version is ready the next time Lantern starts.
+
+### Update server
+
+![lanternautoupdates - server process](https://cloud.githubusercontent.com/assets/385670/6097042/cb08d42c-af72-11e4-9ca4-d09af2fbb11b.png)
+
+The update server holds a list of releases and waits for queries from clients.
+Clients will send their own checksum and the server will compare that checksum
+against the checksum of the latest release, if they don't match a binary diff
+will be generated. This binary diff can be used by the client to patch itself.
+
+The update server may or may not be used as a download server. Clients will
+pull binary diffs from this location, the actual patch's URL will be provided
+by the update server.
+
+### Lantern client
+
+![lanternautoupdates - auto update process](https://cloud.githubusercontent.com/assets/385670/6097031/755f89c6-af72-11e4-82ea-0c82f27160b2.png)
+
+A client will compute the checksum of its executable file and send it to an
+update server periodically. When the update server replies with a special
+message meaning that a new version is available, the client will download the
+binary patch, apply it to a temporary file and check the signature, if the
+signature is what the client expects, the original executable will be replaced
+with the patched one.
```

Comment:
Can you add a section on "Security Considerations" addressing things such as the potential for malware overwriting lantern.exe to get through the Windows firewall?